pkg/webhooks/webhookpatch.go

```diff
@@ -97,53 +93,49 @@ func PatchCertLoop(injectionWebhookConfigName, webhookName, caBundlePath string,
 				if oldConfig.ResourceVersion != newConfig.ResourceVersion {
 					for i, w := range newConfig.Webhooks {
 						if w.Name == webhookName && !bytes.Equal(newConfig.Webhooks[i].ClientConfig.CABundle, caCertPem) {
-							log.Infof("Detected a change in CABundle, patching MutatingWebhookConfiguration again")
-							shouldPatch <- struct{}{}
+							log.Infof("Detected a change in CA bundle, patching MutatingWebhookConfiguration for %s", newConfig.Name)
+							go doPatchWithRetries(client, revision, newConfig.Name, webhookName, caCertPem)
 							break
 						}
 					}
 				}
 			},
+			AddFunc: func(obj interface{}) {
+				config := obj.(*v1beta1.MutatingWebhookConfiguration)
+				log.Infof("New webhook config added, patching MutatingWebhookConfiguration for %s", config.Name)
+				go doPatchWithRetries(client, revision, config.Name, webhookName, caCertPem)
+			},
 		},
 	)
 	go controller.Run(stopCh)
+}
 
-	go func() {
-		var delayedRetryC <-chan time.Time
-		if retry {
-			delayedRetryC = time.After(delayedRetryTime)
-		}
-
-		for {
-			select {
-			case <-delayedRetryC:
-				if retry := doPatch(client, injectionWebhookConfigName, webhookName, caCertPem); retry {
-					delayedRetryC = time.After(delayedRetryTime)
-				} else {
-					log.Infof("Retried patch succeeded")
-					delayedRetryC = nil
-				}
-			case <-shouldPatch:
-				if retry := doPatch(client, injectionWebhookConfigName, webhookName, caCertPem); retry {
-					if delayedRetryC == nil {
-						delayedRetryC = time.After(delayedRetryTime)
-					}
-				} else {
-					delayedRetryC = nil
-				}
+func doPatchWithRetries(client kubernetes.Interface, revision, webhookConfigName, webhookName string, caCertPem []byte) {
```

Comment:
One concern I have is if we are somehow failing to patch, we may trigger this multiple times. I can imagine some worst case scenario where two instances are writing different cabundles, continually triggering the update, which fires off a goroutine and retries infinitely, and causing an infinite loop of more and more goroutines contributing to this.

I think the old logic handled this by essentially serializing it, or we can use a queue like some other controllers do

Reply:
Restructured the controller to use a task queue, think this should be addressed in the latest version